e2e/node/basic/syncTime.test.ts

```diff
@@ -0,0 +1,496 @@
+      expect(actorResponse).toEqual(greetRes);
+
+      expect(mockReplica.getV3CallSpy(canisterId.toString())).toHaveBeenCalledTimes(1);
+      expectV3CallRequest(
+        mockReplica.getV3CallReq(canisterId.toString(), 0),
+        {
+          nonce,
+          sender,
+          pubKey: new Uint8Array(identity.getPublicKey().toDer()),
+          signature,
+        },
+        'V3 call body',
+      );
+    });
+
+    it('should sync time when the local time does not match the subnet time', async () => {
+      const agent = await HttpAgent.create({
+        host: mockReplica.address,
+        rootKey: keyPair.publicKeyDer,
+        identity,
+      });
+      const actor = createCounterActor(canisterId, agent);
+      const sender = identity.getPrincipal();
+
+      mockReplica.setV3CallSpyImplOnce(canisterId.toString(), (_req, res) => {
+        res.status(400).send(new TextEncoder().encode(INVALID_EXPIRY_ERROR));
+      });
+
+      const { responseBody: readStateResponse } = await prepareV2ReadStateTimeResponse({
+        keyPair,
+      });
+      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
+        res.status(200).send(readStateResponse);
+      });
+      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
+        res.status(200).send(readStateResponse);
+      });
+      mockReplica.setV2ReadStateSpyImplOnce(canisterId.toString(), (_req, res) => {
+        res.status(200).send(readStateResponse);
+      });
+
+      const { responseBody: callResponse, requestId } = await prepareV3Response({
+        canisterId,
+        methodName: greetMethodName,
+        arg: greetArgs,
+        sender,
+        reply: greetReply,
+        keyPair,
+        date,
+        nonce,
+      });
+      const signature = await identity.sign(concat(IC_REQUEST_DOMAIN_SEPARATOR, requestId));
+      mockReplica.setV3CallSpyImplOnce(canisterId.toString(), (_req, res) => {
+        res.status(200).send(callResponse);
+      });
+
+      const actorResponse = await actor.greet.withOptions({ nonce })(greetReq);
+
+      expect(actorResponse).toEqual(greetRes);
+      expect(mockReplica.getV3CallSpy(canisterId.toString())).toHaveBeenCalledTimes(2);
+
+      const req = mockReplica.getV3CallReq(canisterId.toString(), 0);
+      expectV3CallRequest(
+        req,
+        {
+          nonce,
+          sender,
+          pubKey: new Uint8Array(identity.getPublicKey().toDer()),
+          signature,
+        },
+        'V3 call body',
+      );
+
+      const reqTwo = mockReplica.getV3CallReq(canisterId.toString(), 1);
+      expect(reqTwo).toEqual(req);
+
+      expect(mockReplica.getV2ReadStateSpy(canisterId.toString())).toHaveBeenCalledTimes(3);
+      expectV2ReadStateRequest(
+        mockReplica.getV2ReadStateReq(canisterId.toString(), 0),
+        {
+          sender: anonIdentity.getPrincipal(),
+        },
+        'V3 read state body one',
```

Comment:
Really not important, but makes the code more readable
```suggestion
        'V2 read state body one',
```